### PR TITLE
Issue 21518 date time field changes for all language versions in url mapped content

### LINKED ISSUE #21518: Date/Time Field Changes for All Language Versions in URL Mapped Content
**Describe the bug**
When dealing with multi-lingual URL mapped content, when the `date/time` field is updated for one language version it is then updated for all other language versions too. 

**To Reproduce**
1. Go to demo.dotcms.com
2. Create a Spanish version of a blog content type
3. Change the posting date of the english version
4. See the Spanish version populate that same date automatically 

**Expected behavior**
The date.time field would be populated on the inode and not the identifier to allow for distinct `date/time` values across language versions.  Another possibility would be to have multiple dates available and be able to set which one would be served as the `published` or `posting` date

**Screenshots**
Possible solution if you can set "Another Date" Field in Screenshot

Screen Shot 2022-01-12 at 11 15 31 AM

**Desktop (please complete the following information):**
 - Versions : 22, 21.06.4 

**Additional context**
As reported by a customer: https://dotcms.zendesk.com/agent/tickets/106208

### PR DESCRIPTION
**Pull request by freddyucv**
We had a` syspublish_date` and `sysexpire_date` field in the Identifier table to store the publish/expire date values, we always took the values from this fields so all the contentlet's versions were going to have the values all the time/

Now we want to have to behavior for this:

- If 'uniquePublishExpireDate' property is set to false (default) we want that each Contentlet's version to have its own publish/expire value.
- If 'uniquePublishExpireDate' property is set to true we want that all the Contentlet's version to have the same publish/expire value, I mean take the value from the Identifier.

Also:

- We are taking the allow format date value to ES from the es-content-mapping.json file when create  a newly index, but also we are using this value here:

https://github.com/dotCMS/core/blob/master/dotCMS/src/main/java/com/dotcms/content/elasticsearch/util/ESMappingUtilHelper.java#L372

so we are duplicating the format value, so now with this change we are taking the value from the es-content-mapping.json in both places.

- I created a new Util class

https://github.com/dotCMS/core/compare/issue-21518-Date-Time-Field-Changes-for-All-Language-Versions-in-URL-Mapped-Content?expand=1#diff-9129120d9ade789b78924212ab30d42967a0057c7a88b8e27017929c5a19b1e3R1

### CODE REVIEW

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java [hunk 1]
```diff
@@ -1277,6 +1274,242 @@ public void flushVanityURLCacheRight() throws ServletException, IOException {
...
+     * Should: Each {@link Contentlet} be saved and index with the right publish date
+     *
+     * @throws DotDataException
+     * @throws DotSecurityException
+     */
+    @Test
+    public void differentVersionWithDifferentPublishDate() throws DotDataException, DotSecurityException {
+
+        final Language language1 = new LanguageDataGen().nextPersisted();
+        final Language language2 = new LanguageDataGen().nextPersisted();
```
Comment (jcastro-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java:1293]: Shouldn't we delete all the test languages in the test methods after they run? It seems like we'll end up with several tests languages, which might affect other tests in the end.

File: dotCMS/src/main/java/com/dotmarketing/portlets/contentlet/model/Contentlet.java [hunk 2]
```diff
@@ -752,7 +753,7 @@ public void setDateProperty(com.dotcms.contenttype.model.field.Field field, Date
 	 */
 	public Date getDateProperty(String fieldVarName) throws DotRuntimeException {
 		try{
-			return (Date)map.get(fieldVarName);
+			return map.containsKey(fieldVarName) ? (Date) map.get(fieldVarName) : null;
 		}catch (Exception e) {
 			 throw new DotRuntimeException("Unable to retrive field value", e);
```
Comment (jcastro-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/contentlet/model/Contentlet.java:758]: Not related to your specific changes, but can you add the name of the Date field to the error message please?

**Review by jcastro-dotcms — CHANGES_REQUESTED** [review]

File: dotCMS/src/integration-test/java/com/dotmarketing/portlets/contentlet/business/ContentletAPITest.java [hunk 3]
```diff
@@ -3816,117 +3819,122 @@ public void saveContentWithExistingIdentifier() throws Exception {
...
+
+            final long count = APILocator.getContentletAPI().indexCount(q, user, false);
+            assertEquals(1, count);
+        } finally {
+            ContentletTransformer.setUniquePublishExpireDatePerLanguages(uniquePublishExpireDatePerLanguages);
+        }
     }
 
-    @Test
+    ////@Test
```
Comment (dsilvam) [inline dotCMS/src/integration-test/java/com/dotmarketing/portlets/contentlet/business/ContentletAPITest.java:3937]: should we delete this test? if not, please explain why commenting it out  @freddyucv

**Review by dsilvam — APPROVED** [review]

File: dotCMS/src/main/java/com/dotmarketing/portlets/contentlet/model/Contentlet.java (hunk 2 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotmarketing/portlets/contentlet/model/Contentlet.java]: done https://github.com/dotCMS/core/pull/22096/commits/fc2bbe074dc78a05ae755a49843d0b9d4662702a#diff-1367236d40a612b923ececfa521053143f026f60f035c9f9963aae4efdfaf034R758

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/util/ESMappingUtilHelper.java [hunk 4]
```diff
@@ -366,29 +367,39 @@ private Optional<List<Tuple2<String, JSONObject>>> getMappingForField(final Fiel
...
+                mappingForField = "{\n\"type\":\"date\",\n";
+
+                try {
+                    final Map<String, Object> jsonFileContent = JsonUtil.getJsonFileContent(
+                            "es-content-mapping.json");
+
+                    mappingForField += String.format("\"format\": \"%s\"\n}",
+                            ((List) jsonFileContent.get("dynamic_date_formats")).get(0));
+                } catch (IOException e) {
+                    throw new JSONException(e);
```
[the remark this answers is not in the document]
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/content/elasticsearch/util/ESMappingUtilHelper.java:383]: done https://github.com/dotCMS/core/pull/22096/commits/fc2bbe074dc78a05ae755a49843d0b9d4662702a#diff-8210ce6c49e98569dc879cd2c50b9d4ff30861f3285c42f7b0a2b16129d2b5dbR382

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java (hunk 1 above)
Comment (freddyucv) [reply dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java:1293]: done https://github.com/dotCMS/core/pull/22096/commits/9f0e7869e66bb92bf31604b93a882ffc887acd55#diff-093de07a53a725b587963fd60c024f52e089c4f78e7ea88ba8e3a3a0cdaecab9R1330

**Review by jcastro-dotcms — APPROVED** [review]

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java [hunk 5]
```diff
@@ -1277,6 +1274,259 @@ public void flushVanityURLCacheRight() throws ServletException, IOException {
...
+                    .nextPersisted();
+
+            contentlet2 = ContentletDataGen.checkout(contentlet1);
+            contentlet2.setLanguageId(language2.getId());
+            contentlet2.setProperty(publishField.variable(), afterTomorrow);
+            ContentletDataGen.checkin(contentlet2);
+
+            checkFromElasticSearch(publishField, tomorrow, afterTomorrow, contentlet1);
+
+            checkFromDataBase(publishField, tomorrow, afterTomorrow, contentlet1);
```
Comment (nollymar) [inline dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java:1328]: why contentlet2 result is never checked out?

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java [hunk 6]
```diff
@@ -1277,6 +1274,259 @@ public void flushVanityURLCacheRight() throws ServletException, IOException {
...
+                    .setProperty(publishField.variable(), tomorrow)
+                    .languageId(language1.getId())
+                    .nextPersisted();
+
+            final Contentlet contentlet2 = ContentletDataGen.checkout(contentlet1);
+            contentlet2.setLanguageId(language2.getId());
+            contentlet2.setProperty(publishField.variable(), afterTomorrow);
+            ContentletDataGen.checkin(contentlet2);
+
+            checkFromElasticSearch(publishField, afterTomorrow, afterTomorrow, contentlet1);
```
Comment (nollymar) [inline dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java:1388]: I don't get why the publish date for content2 is never verified

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java [hunk 7]
```diff
@@ -1277,6 +1274,259 @@ public void flushVanityURLCacheRight() throws ServletException, IOException {
...
+            checkFromDataBase(publishField, afterTomorrow, afterTomorrow, contentlet1);
+        }finally {
+            ContentletTransformer.setUniquePublishExpireDatePerLanguages(uniquePublishExpireDate);
+        }
+    }
+
+    private void checkFromDataBase(Field publishField, Date tomorrow, Date afterTomorrow,
```
Comment (nollymar) [inline dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java:1395]: please add some javadoc here to understand better what the method does

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java [hunk 8]
```diff
@@ -1277,6 +1274,259 @@ public void flushVanityURLCacheRight() throws ServletException, IOException {
...
+                assertEquals(tomorrow,
+                        ((Contentlet) versionable).getDateProperty(publishField.variable()));
+            } else {
+                assertEquals(afterTomorrow,
+                        ((Contentlet) versionable).getDateProperty(publishField.variable()));
+            }
+        }
+    }
+
+    private void checkFromElasticSearch(final Field field, final Date date1, final Date date2,
```
Comment (nollymar) [inline dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java:1412]: Add some doc

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java (hunk 5 above)
Comment (freddyucv) [reply dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java:1328]: really the method is checking both

https://github.com/dotCMS/core/blob/9f0e7869e66bb92bf31604b93a882ffc887acd55/dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java#L1412

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java (hunk 6 above)
Comment (freddyucv) [reply dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java:1388]: The methods are checking both versiona

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java (hunk 7 above)
Comment (freddyucv) [reply dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java]: done https://github.com/dotCMS/core/pull/22096/commits/e0c5b1472ea4d41687ac2e206517d9545bbe6bc5

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java (hunk 5 above)
Comment (dsilvam) [reply dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java:1327]: I also found it confusing @nollymar . I think the `checkFromDataBase` method needs java doc or maybe done it in a different way. Basically the method implementation finds the two versions given only contentl1, hence not needing to pass contentlet2 in the params.

**Pull request merged**